### PR TITLE
Security jwt filter implementation

### PR DESCRIPTION
**Pull request by vetterocki**
Implemented JWT Bearer token filtration + updated existing endpoints to not violate created security rules.
Things to be updated later in this PR:
 --> Update web controllers tests with usage of security
 --> Analyze user (and according to result of discussion remodel endpoints security access) permissions to endpoints

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -54,6 +54,23 @@
...
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-security</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.security</groupId>
+            <artifactId>spring-security-test</artifactId>
+            <scope>test</scope>
+        </dependency>
+
```
Comment (Kivanval) [inline pom.xml:72]: Put versions in the properties

File: src/main/java/com/qunite/api/domain/User.java [hunk 2]
```diff
@@ -117,4 +136,41 @@ public boolean equals(Object o) {
...
+
+  @Override
+  public String getPassword() {
+    return password;
+  }
+
+  @Override
+  public String getUsername() {
+    return username;
+  }
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/domain/User.java:153]: Explicit labeling is not necessary

File: src/main/java/com/qunite/api/domain/User.java [hunk 3]
```diff
@@ -117,4 +136,41 @@ public boolean equals(Object o) {
   public int hashCode() {
     return getClass().hashCode();
   }
+
+  @Override
+  public Collection<? extends GrantedAuthority> getAuthorities() {
+    return Set.of(new SimpleGrantedAuthority(ROLES));
+  }
+
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/domain/User.java:144]: I don't really understand the point of roles

File: src/main/java/com/qunite/api/security/JwtConstants.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package com.qunite.api.security;
+
+
+import lombok.experimental.UtilityClass;
+
+@UtilityClass
+public class JwtConstants {
+  public static final String TOKEN_PREFIX = "Bearer ";
+  public static final String ROLES = "users";
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/security/JwtConstants.java:9]: Why is it plural

File: src/main/java/com/qunite/api/security/JwtConstants.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.qunite.api.security;
+
+
+import lombok.experimental.UtilityClass;
+
+@UtilityClass
+public class JwtConstants {
+  public static final String TOKEN_PREFIX = "Bearer ";
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/security/JwtConstants.java:8]: This is a prefix only for the token bearer, I don't see the point of putting it in a constant

File: src/main/java/com/qunite/api/service/QueueService.java [hunk 6]
```diff
@@ -19,4 +19,11 @@ public interface QueueService {
   Optional<Integer> getMemberPositionInQueue(Long memberId, Long queueId);
 
   void deleteById(Long queueId);
+
+  Optional<Queue> findByUserComparingBoth(Long queueId, String loginData);
+
+  Optional<Queue> findByUserComparingToQueueManagers(Long queueId, String loginData);
+
+  Optional<Queue> findByUserComparingToCreatorId(Long queueId, String loginData);
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/service/QueueService.java:27]: Strange names

File: src/main/java/com/qunite/api/service/QueueServiceImpl.java [hunk 7]
```diff
@@ -64,4 +66,36 @@ public List<Queue> findAll() {
...
+        .flatMap(queue -> queue.getManagers().stream()
+            .map(User::getId)
+            .anyMatch(managerId -> userService.compareUserIdToLoginData(loginData, managerId)
+                .isPresent()) ? Optional.of(queue) : Optional.empty());
+  }
+
+  @Override
+  @Transactional
+  public Optional<Queue> findByUserComparingBoth(Long queueId, String loginData) {
+    return this.findById(queueId)
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/service/QueueServiceImpl.java:93]: Why use this?

File: src/main/java/com/qunite/api/service/UserServiceImpl.java [hunk 8]
```diff
@@ -50,5 +63,35 @@ public Optional<List<Queue>> getManagedQueues(Long userId) {
...
+  @Override
+  @Transactional
+  public User register(String username, String email, String password) {
+    if (!userRepository.existsByUsernameOrEmail(username, email)) {
+      return userRepository.save(new User(username, email, passwordEncoder.encode(password)));
+    } else {
+      throw new ResponseStatusException(HttpStatus.BAD_REQUEST,
+          "User with such credentials already exists!");
+    }
+  }
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/service/UserServiceImpl.java:75]: change to createOne()

File: src/main/java/com/qunite/api/service/UserServiceImpl.java [hunk 9]
```diff
@@ -50,5 +63,35 @@ public Optional<List<Queue>> getManagedQueues(Long userId) {
...
+
+  @Override
+  @Transactional
+  public Optional<User> compareUserIdToLoginData(String loginData, Long id) {
+    return this.findOne(id)
+        .flatMap(founded -> this.findByUsernameOrEmail(loginData)
+            .map(founded::equals)
+            .flatMap(condition -> condition ? Optional.of(founded) : Optional.empty()));
+  }
+
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/service/UserServiceImpl.java:95]: It looks quite strange, a map to boolean

File: src/main/java/com/qunite/api/web/controller/AuthenticationController.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+@RequiredArgsConstructor
+@RequestMapping(path = "/auth", produces = MediaType.APPLICATION_JSON_VALUE)
+@Tag(name = "Authentication controller")
+@RestController
+public class AuthenticationController {
+  private final UserService userService;
+  private final AuthResponseMapper responseMapper;
+
+  @Operation(summary = "Register user", responses = @ApiResponse(responseCode = "200"))
+  @PostMapping("/register")
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/web/controller/AuthenticationController.java:32]: I propose a sign-up name for the endpoint for consistency

File: src/main/java/com/qunite/api/service/UserServiceImpl.java [hunk 11]
```diff
@@ -1,23 +1,31 @@
...
 import org.springframework.stereotype.Service;
 import org.springframework.transaction.annotation.Transactional;
+import org.springframework.web.server.ResponseStatusException;
 
 @Service
 @RequiredArgsConstructor
 public class UserServiceImpl implements UserService {
-
   private final UserRepository userRepository;
+  private final PasswordEncoder passwordEncoder;
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/service/UserServiceImpl.java:22]: Code at the filter level

**Review by Kivanval — CHANGES_REQUESTED** [review]

File: src/main/java/com/qunite/api/security/JwtAuthorizationFilter.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+          var authToken = new UsernamePasswordAuthenticationToken(
+              username, null, Set.of(new SimpleGrantedAuthority(ROLES))
+          );
+          SecurityContextHolder.getContext().setAuthentication(authToken);
+        });
+    filterChain.doFilter(request, response);
+  }
+
+  private boolean isHeaderValid(String authHeader) {
+    return Objects.nonNull(authHeader) && authHeader.startsWith(TOKEN_PREFIX);
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/security/JwtAuthorizationFilter.java:50]: I'm not sure, but nonNull can throw an error if it's null, I recommend using the usual !=null

**Review by Kivanval — CHANGES_REQUESTED** [review]

File: src/main/java/com/qunite/api/web/controller/UserController.java [hunk 13]
```diff
@@ -49,58 +50,66 @@ public ResponseEntity<List<UserDto>> all() {
       @ApiResponse(responseCode = "200"),
       @ApiResponse(responseCode = "404", content = @Content())
   })
-  public ResponseEntity<UserDto> getById(@PathVariable Long id) {
-    return ResponseEntity.of(userService.findOne(id).map(userMapper::toDto));
+  public ResponseEntity<UserDto> getById(Principal principal, @PathVariable Long id) {
+    log.info("hi there");
```
Comment (eternityduck) [inline src/main/java/com/qunite/api/web/controller/UserController.java:54]: Is this necessary?

File: src/main/java/com/qunite/api/service/QueueServiceImpl.java [hunk 14]
```diff
@@ -48,11 +51,28 @@ public Optional<Integer> getMemberPositionInQueue(Long memberId, Long queueId) {
...
   @Transactional
-  public void deleteById(Long queueId) {
-    queueRepository.deleteById(queueId);
+  public void deleteById(Long queueId, String loginData) {
+    var queueById = findById(queueId);
+    if (queueById.isPresent()) {
+      if (isUserByCredentialsQueueCreator(queueId, loginData)) {
+        queueRepository.deleteById(queueId);
+      } else {
+        throw new AccessDeniedException("Credentials does not match the queue creator");
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/service/QueueServiceImpl.java:62]: rename to "The user is not the creator"

File: src/main/java/com/qunite/api/security/JwtService.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+import java.util.Optional;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+@Component
+public class JwtService {
+  @Value("${jwt.secret}")
+  private String secret;
+
+  @Value("${jwt.issuer}")
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/security/JwtService.java:17]: use spring.application.name

File: src/main/java/com/qunite/api/data/UserRepository.java [hunk 16]
```diff
@@ -1,9 +1,18 @@
...
 import org.springframework.stereotype.Repository;
 
 @Repository
 public interface UserRepository extends JpaRepository<User, Long> {
+  @Query("SELECT u FROM User u WHERE u.email = :data OR u.username = :data")
+  Optional<User> findByEmailOrUsername(@Param("data") String loginData);
+
+  boolean existsByUsernameOrEmail(String username, String email);
+
+
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/data/UserRepository.java:17]: unnecessary indents

File: src/main/java/com/qunite/api/web/controller/UserController.java [hunk 17]
```diff
@@ -76,31 +76,33 @@ public ResponseEntity<List<QueueDto>> getCreatedQueues(@PathVariable Long id) {
...
-  @DeleteMapping("/{id}")
-  @Operation(summary = "Delete user by id", responses = @ApiResponse(responseCode = "204"))
-  public ResponseEntity<Void> deleteById(@PathVariable Long id) {
-    userService.deleteOne(id);
-    return ResponseEntity.noContent().build();
-  }
-
-  @PatchMapping("/{id}")
-  @Operation(summary = "Update user by id", responses = {
+  @PatchMapping("/")
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/web/controller/UserController.java:85]: /self

File: src/main/java/com/qunite/api/web/mapper/AuthResponseMapper.java [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+import com.auth0.jwt.interfaces.DecodedJWT;
+import com.qunite.api.web.dto.auth.AuthenticationResponse;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+import org.mapstruct.ReportingPolicy;
+
+@Mapper(unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface AuthResponseMapper {
+
+  @Mapping(source = "token", target = "accessToken")
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/web/mapper/AuthResponseMapper.java:12]: just use token field name

File: src/main/java/com/qunite/api/web/controller/UserController.java [hunk 19]
```diff
@@ -76,31 +76,33 @@ public ResponseEntity<List<QueueDto>> getCreatedQueues(@PathVariable Long id) {
...
-        .map(user -> userMapper.partialUpdate(userUpdateDto, user))
+    return ResponseEntity.of(userService.findByUsernameOrEmail(principal.getName())
+        .map(founded -> userMapper.partialUpdate(userUpdateDto, founded))
         .map(userService::createOne)
-        .map(userMapper::toDto)
-    );
+        .map(userMapper::toDto));
+  }
+
+  @DeleteMapping("/")
```
Comment (Kivanval) [inline src/main/java/com/qunite/api/web/controller/UserController.java:98]: /self

**Review by Kivanval — CHANGES_REQUESTED** [review]

**Review by Kivanval — APPROVED** [review]

File: src/main/java/com/qunite/api/web/controller/UserController.java [hunk 20]
```diff
@@ -76,31 +72,33 @@ public ResponseEntity<List<QueueDto>> getCreatedQueues(@PathVariable Long id) {
                 .map(queueMapper::toDto).toList()));
   }
 
-  @PostMapping
-  @Operation(summary = "Create user", responses = @ApiResponse(responseCode = "201"))
-  public ResponseEntity<UserDto> createUser(@Valid @RequestBody UserCreationDto userCreationDTO) {
-    var created = userService.createOne(userMapper.toEntity(userCreationDTO));
-    return new ResponseEntity<>(userMapper.toDto(created), HttpStatus.CREATED);
+  @GetMapping("/self")
+  public ResponseEntity<UserDto> getSelf(Principal principal) {
```
Comment (xd6d) [inline src/main/java/com/qunite/api/web/controller/UserController.java:76]: missing swagger description

**Review by xd6d — APPROVED** [review]

**Pull request merged**